### ISSUE TITLE
Documentation of .mdlDowngrade_ method in wiki is vague

### ISSUE THREAD

**Issue by leifoolsen** [google/material-design-lite#4209]
Component (if any): 

What are you trying to do or find out more about?
Clean up the component if it needs to be removed.

Where have you looked?
https://github.com/google/material-design-lite/wiki/Making-your-first-JS-component

Where did you expect to find this information?
https://github.com/google/material-design-lite/wiki/Making-your-first-JS-component

From wiki:
"Since this component adds an event listener to the element the mdlDowngrade_ method should be implemented to clean up the component if it needs to be removed."

```javascript
Dropdown.prototype.mdlDowngrade_ = function() {
  'use strict';
  this.element_.removeEventListener('click', this.boundClickHandler);
};
```
If you don't add a listener to the **mdl-componentdowngraded** event in the `init` method, the `mdlDowngrade_` method will not be called. So the `init` method should have code similar to this - I think.

```javascript
Dropdown.prototype.init = function() {
    "use strict";
    this.boundClickHandler = this.clickHandler.bind(this);
    this.element_.addEventListener('click', this.boundClickHandler);

    this.element_.addEventListener('mdl-componentdowngraded', this.mdlDowngrade_.bind(this.element_));
};
``` 

And in mdlDowngrade_
```javascript
Dropdown.prototype.mdlDowngrade_ = function() {
  'use strict';
  this.element_.removeEventListener('click', this.boundClickHandler);
  this.element_.removeEventListener('mdl-componentdowngraded', this);
  this.classList.remove(IS_UPGRADED);
};
```

A (pseudo) test 
```javascript
it('downgrades successfully', () => {

  // Code to insert component via insertAdjacentHTML
  ........
   
  // Upgrade
  const element = document.queryselector('#my-dropdown');
  componentHandler.upgradeElement(element, 'Dropdown');

  // Downgrade
  componentHandler.downgradeElements(element);
  expect(element.getAttribute('data-upgraded')).to.not.include('Dropdown');
});
```

Regards
Leif Olsen

**Comment by leifoolsen** [comment]
I have upgraded MDL to version 1.1.3 in my project - and now the **mdl-componentdowngraded** does not fire!

**Comment by leifoolsen** [comment]
Codechange in componentHandler could be an issue:

1.1.2
```javascript
function deconstructComponentInternal(component) {
  if (component) {
    var componentIndex = createdComponents_.indexOf(component);
    createdComponents_.splice(componentIndex, 1);

    var upgrades = component.element_.getAttribute('data-upgraded').split(',');
    var componentPlace = upgrades.indexOf(component[componentConfigProperty_].classAsString);
    upgrades.splice(componentPlace, 1);
    component.element_.setAttribute('data-upgraded', upgrades.join(','));

    var ev = document.createEvent('Events');
    ev.initEvent('mdl-componentdowngraded', true, true);
    component.element_.dispatchEvent(ev);
  }
}
```

1.1.3
```javascript
  function deconstructComponentInternal(component) {
    if (component) {
      var componentIndex = createdComponents_.indexOf(component);
      createdComponents_.splice(componentIndex, 1);

      var upgrades = component.element_.getAttribute('data-upgraded').split(',');
      var componentPlace = upgrades.indexOf(component[componentConfigProperty_].classAsString);
      upgrades.splice(componentPlace, 1);
      component.element_.setAttribute('data-upgraded', upgrades.join(','));

      var ev;
      if ('CustomEvent' in window && typeof window.CustomEvent === 'function') {
        ev = new Event('mdl-componentdowngraded', {
          'bubbles': true, 'cancelable': false
        });
      } else {
        ev = document.createEvent('Events');
        ev.initEvent('mdl-componentdowngraded', true, true);
      }
    }
  }
```

**Comment by leifoolsen** [comment]
I think code for custom event shold be:

```javascript
      const evt = new CustomEvent('mdl-componentdowngraded', {
        bubbles: true,
        cancelable: false,
        detail: { }
      });
      component.element_.dispatchEvent(evt);
```

**Comment by dgrubelic** [comment]
That `CustomEvent` won't work in IE10 which is among supported browsers for v1.

**Comment by leifoolsen** [comment]
According to http://caniuse.com/#feat=customevent there should be no need to polyfill CustomEvent.

This line is missing i 1.1.3:
component.element_.dispatchEvent(evt);

**Comment by dgrubelic** [comment]
Yeah, you can use `document.createEvent('CustomEvent')`, not `new CustomEvent()`. That is what i'm saying...

**Comment by leifoolsen** [comment]
Anyway, no event is fired due to missing code :)

**Comment by leifoolsen** [comment]
This is the event polyfill from Google Dialog polyfill:
```javascript
  var supportCustomEvent = window.CustomEvent;
  if (!supportCustomEvent || typeof supportCustomEvent == 'object') {
    supportCustomEvent = function CustomEvent(event, x) {
      x = x || {};
      var ev = document.createEvent('CustomEvent');
      ev.initCustomEvent(event, !!x.bubbles, !!x.cancelable, x.detail || null);
      return ev;
    };
    supportCustomEvent.prototype = window.Event.prototype;
  }
```

**Comment by dgrubelic** [comment]
Yeah, that should a part of MDL component handler script or something so it is globally available so we don't have to check it every time in every component.

**Comment by leifoolsen** [comment]
... but the event should be fired? So I think it's a bug in code.

**Comment by dgrubelic** [comment]
No, i agree definitely. Just saying on the side topic. So we can kill two birds with one stone :smile_cat:

**Comment by leifoolsen** [comment]
@dgrubelic:  A short, related, question.
How do you handle downgrading of internal components, e.g. downgrading of the **MaterialTextfield** component? I can not find any code that actually detaches the attached event handlers of this component.

**Comment by leifoolsen** [comment]
BTW: I have a mdl selectfield here: https://github.com/leifoolsen/mdl-ext/tree/master/src/selectfield

**Comment by dgrubelic** [comment]
I think that `MaterialTextfield` component is missing mdlDowngrade_() method...so i don't. :)

Yes, this is plain html select only wrapped in `mdl-selectfield` div. What i did was entirely new component what will look the same in every browser on every device. With accessibility included in component.

For example, what happens when you set select to be multiple select? How do you handle that to look like Material design component? (we can continue this on my repo for example)

**Comment by leifoolsen** [comment]
I would personally avoid using a multi select bc. that would normally be a list - which will take up too much space in responsive design. I think a `<datalist>´based on e.g. [Awesomplete](http://leaverou.github.io/awesomplete/) would be a better solution for multiselect. See: http://leaverou.github.io/awesomplete/#advanced-examples

**Comment by Garbee** [comment]
From what I can make out, you are trying to figure out why the downgrade method isn't called anymore.

The documentation is outdated and we made a technical BC break in 1.1 to not need the downgrade method anymore. Basically it works like this...

Downgrade away, it removes the internal component reference. Then just remove the DOM nodes. Done. The garbage collection in the engine will kick-in and remove the event listeners (assuming you don't have references to those nodes in *your* code remaining.)

The downgrade method was a hack because of me not running a proper test environment to check for event listener removal. So, I thought engines didn't clean up when in fact they do. Once we realized that isn't the case the downgrade method didn't make much sense to keep around. It doesn't hurt external components that *have* one, it is just dead code that needs to be removed later on. But now all components have a way to remove the handlers internal reference, via the downgrade method of the handler itself.

**Comment by leifoolsen** [comment]
I can see that removing listeners is no longer needed for modern browsers. But there could be other stuff that need cleanup - and the only way to do do this is trough a listener. Also according to this [Stack Overflow q/a](http://stackoverflow.com/questions/12528049/if-a-dom-element-is-removed-are-its-listeners-also-removed-from-memory) there can still be issues with modern browsers.

**Comment by Garbee** [comment]
The only caveat is as I stated "If you have any references in your code". Which means you should do cleanup on your own to remove those references (therefore completely destroying the node instead of having it kept in memory in a non-attached state.)

The components *should* only add *internal* event listeners and modify structure internal to their root reference (in almost all cases.) This means simply destroying the root node (and therefore all children) from the DOM is enough if your application doesn't keep lingering references.

**Comment by leifoolsen** [comment]
If the downgrade method should not be called anymore, then the initialization code for event emitter sholud be removed as well.  This code is redundant - since no event should be emitted:

```javascript
      var ev;
      if ('CustomEvent' in window && typeof window.CustomEvent === 'function') {
        ev = new Event('mdl-componentdowngraded', {
          'bubbles': true, 'cancelable': false
        });
      } else {
        ev = document.createEvent('Events');
        ev.initEvent('mdl-componentdowngraded', true, true);
      }
```

**Comment by leifoolsen** [comment]
... also the class `is-upgraded` still remains in the classlist after a call to `componentHandler.downgradeElements(element)`. Shouldn't the downgrade method remove this class as well?

**Comment by Garbee** [comment]
Why are you trying to remove that class specifically? The *only* reason you should be downgrading is to remove something from the DOM completely. Any other cases were not considered with 1.x so they aren't supported.

The event will remain as well. Applications expecting it to fire for some reason may still have functionality dependent on a downgrade happening (say analytics events.) That would be a massively breaking change/removal. Something still happens (internal reference removal) so it is "downgraded" in a way, making it safe for deletion. Yes, event name isn't semantic... but it would be a BC break to rename it.

**Comment by leifoolsen** [comment]
@Garbee: Thanks for a good and thorough explanation of this topic. The confusion with me was when I read the documentation regarding this. The conclusion is that I do not need to do anything in connection with the downgrading, just remove nodes.

**Comment by leifoolsen** [comment]
@Garbee: Excuse me for bringing this subject up again, but for components converted to ES6 classes in MDL v2, e.g. MaterialCheckbox, event listeners are removed explicitly. The conclusion then must be that if one develops components for v2, then the event listener should be removed?

Another small improvement that can be done (hopefully I'm not too picky now). The components.js, initComponents method could be rewritten to something like this, using the spread operator:

```javascript
  static initComponents(docRoot) {
    [...docRoot.querySelectorAll(`.${this.classes_.ROOT}.${this.classes_.JS}`)].forEach( node => {     
       node[this.strings_.CLASS_NAME] = new this(node);
    });
  }
```
